File: Python/Lab 6/Lab_6.py
from math import factorial, fabs

def init():  # Функція для введення поаткових значень а, b та точності 
    global a 
    a = float(input("Enter a value of a: "))
    global b
    b = float(input("Enter a value of b: "))
    global accuracy 
    accuracy = float(input("Enter an accuracy: "))

def t_cos(x, k):  # Функція яка рахує косинус за допомогою формули Тейлора із заданою k кількістю членів послідовності
    sum = 1
    for i in range(1, k+1):
        sum += pow(-1, i) * (pow(x, 2 * i)) / factorial(2 * i)
    return sum

def calculate(a, b, accuracy):  # Функція для обрахування значення виразу із заданою тоністю
    k = 1                       # Змінна яка задає кількість членів послідовності із ряду Тейлора для обрахування заданої точності
    previous = 0                # Попереднє значення виразу
    current = t_cos(a, k) + pow(t_cos(b, k), 2) + t_cos(a + b, k)         # Розрахування значення виразу
    differ = fabs(current - fabs(previous))                               # Модуль різниці між поточним і попереднім значенням виразу
    print(" {}) Current difference: {}; Current value: {} and previous value {};\n".format(k, differ, current, previous))

    while accuracy <= differ:   # Цикл який перевіряє умову точності
        previous = current
        k += 1
        current = t_cos(a, k) + pow(t_cos(b, k), 2) + t_cos(a + b, k)
        differ = fabs(current - fabs(previous))
        print(" {}) Current difference: {}; Current value: {} and previous value {};\n".format(k, differ, current, previous))
    return previous

init()
# Задання початкових даних

result = calculate(a, b, accuracy)
# Обрахунок виразу із заданою точністю і параметрами

print(" cos(%1.2f) + cos(%1.2f)^2 + cos(%1.2f) = %.10f\n"%(a, b, a + b, result))
# Виведення результатів



    

File: C++/Lab_6/Lab_6.cpp
#include <iostream>
#include <math.h>

using namespace std;

double a, b, accuracy;  // Початкові значення

void init();                                // Функція для введення поаткових значень а, b та точності
double calculate(double, double, double);   // Функція для обрахування значення виразу із заданою тоністю
void result(double finish);                 // Виведення результату
double t_cos(float, int);                   // Функція яка рахує косинус за допомогою формули Тейлора із заданою k кількістю членів послідовності
int fact(int);                              // Факторіал

int main() {

	init();
	long double res = calculate(a, b, accuracy);
	result(res);  // Виведення результату
	return 0;
}

void init()  // Функція для введення поаткових значень а, b та точності
{
	cout << "Enter a value of a, b: "; cin >> a >> b;
	cout << "Enter an accuracy: "; cin >> accuracy;
}

double calculate(double a, double b, double accuracy)  // Функція для обрахування значення виразу із заданою тоністю
{
	long double current = 0;       // Поточне значення виразу
	long double previous, differ;  // Попереднє значення виразу та одуль різниці між поточним і попереднім значенням виразу
	int k = 1;                     // Змінна яка задає кількість членів послідовності із ряду Тейлора для обрахування заданої точності
	do
	{
		previous = current;
		current = t_cos(a, k) + pow(t_cos(b, k), 2) + t_cos(a + b, k);
		differ = fabs(current - fabs(previous));
		printf_s(" %d) Current difference: %.10f; Current value: %f and previous value %f;\n\n", k, differ, current, previous);
		k++;
	} while (accuracy <= differ);   // Цикл який перевіряє умову точності
	return previous;
}

void result(double finish) 
{
	printf_s(" cos(%1.2f) + cos(%1.2f)^2 + cos(%1.2f) = %.10f\n", a, b, a + b, finish);
}

double t_cos(float x, int k)  // Функція яка рахує косинус за допомогою формули Тейлора із заданою k кількістю членів послідовності
{
	long double sum = 1;
	for (int i = 1; i <= k; ++i) {
		sum += pow(-1, i) * (pow(x, 2 * i)) / fact(2 * i);
	}
	return sum;
}

int fact(int n)  // Розрахунок факторіалу ітераційним способом
{
	long sum = 1;
	while (n > 0)
	{
		sum *= n;
		n--;
	}
	return sum;
}